src/main/java/com/aws/greengrass/mqttbridge/clients/IoTCoreClient.java

```diff
@@ -128,22 +163,45 @@ public synchronized void updateSubscriptions(Set<String> topics, @NonNull Consum
         Set<String> topicsToSubscribe = new HashSet<>(topics);
         topicsToSubscribe.removeAll(subscribedIotCoreTopics);
 
-        topicsToSubscribe.forEach(s -> {
-            try {
-                subscribeToIotCore(s);
-                LOGGER.atDebug().kv(TOPIC, s).log("Subscribed to topic");
-                subscribedIotCoreTopics.add(s);
-            } catch (InterruptedException | ExecutionException | TimeoutException e) {
-                LOGGER.atError().kv(TOPIC, s).log("Failed to subscribe");
-            }
-        });
+        subscribeFuture = executorService.submit(() -> subscribeToTopicsWithRetry(topicsToSubscribe));
     }
 
     @Override
     public boolean supportsTopicFilters() {
         return true;
     }
 
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.PreserveStackTrace", "PMD.ExceptionAsFlowControl"})
+    private void subscribeToTopicsWithRetry(Set<String> topics) {
+        // retry only if client is connected; skip if offline.
+        // topics left here should be subscribed when the client is back online (onConnectionResumed event)
+        topics.forEach(s -> {
+            try {
+                RetryUtils.runWithRetry(subscribeRetryConfig, () -> {
+                    try {
+                        if (iotMqttClient.connected()) {
+                            subscribeToIotCore(s);
+                            subscribedIotCoreTopics.add(s);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **184**. At this line the following field is not protected by synchronization: `subscribedIotCoreTopics`. Your code typically protects this field with synchronization on object: `IoTCoreClient.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `IoTCoreClient.this`. For example, like at lines: **107**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **107**, **113**, and **156**.

Reply:
protected with `synchronized`